File: examples/dekita/.claude/hooks/environment-integrity-check.py
#!/usr/bin/env python3
"""フック環境の整合性をチェックする。

Why:
    settings.jsonに登録されたフックファイルが存在しないと、
    他のフックが期待通りに動作しない。セッション開始時に
    不足を検出し、早期に問題を認識させる。

What:
    - settings.jsonから登録済みスクリプトパスを抽出
    - 各ファイルの存在を確認
    - 不足ファイルがあれば復旧手順を含む警告を表示

Remarks:
    - ブロックはせず警告のみ（他の作業は継続可能）
"""

import json
import re
import sys
from pathlib import Path

# Add hooks directory to path for imports
HOOKS_DIR = Path(__file__).parent
sys.path.insert(0, str(HOOKS_DIR))

from lib.execution import log_hook_execution
from lib.session import parse_hook_input

PROJECT_DIR = HOOKS_DIR.parent.parent
SETTINGS_FILE = HOOKS_DIR.parent / "settings.json"


def get_registered_scripts() -> list[str]:
    """Extract all script paths from settings.json (relative to project root)."""
    if not SETTINGS_FILE.exists():
        return []

    try:
        with open(SETTINGS_FILE) as f:
            settings = json.load(f)
    except (OSError, json.JSONDecodeError):
        return []

    hooks = settings.get("hooks", {})
    script_paths: list[str] = []

    # Pattern to match paths after $CLAUDE_PROJECT_DIR (quoted or unquoted)
    # Matches: "$CLAUDE_PROJECT_DIR"/path or $CLAUDE_PROJECT_DIR/path
    pattern = re.compile(
        r'"\$CLAUDE_PROJECT_DIR"/?([^\s"]+)|(?<!["])\$CLAUDE_PROJECT_DIR/?([^\s"]+)'
    )

    for hook_type in hooks.values():
        if isinstance(hook_type, list):
            for entry in hook_type:
                for hook in entry.get("hooks", []):
                    cmd = hook.get("command", "")
                    if not cmd:
                        continue

                    # Find all script paths in the command
                    for match in pattern.finditer(cmd):
                        path = match.group(1) or match.group(2)
                        if path:
                            # Clean up the path (remove trailing quotes, etc.)
                            path = path.strip('"').strip()
                            # Only include actual script files (not directories)
                            if path.endswith((".py", ".sh")):
                                script_paths.append(path)

    return list(set(script_paths))


def check_script_files() -> tuple[list[str], list[str]]:
    """Check which script files exist and which are missing."""
    registered = get_registered_scripts()
    missing: list[str] = []
    found: list[str] = []

    for script_path in registered:
        full_path = PROJECT_DIR / script_path
        if full_path.exists():
            found.append(script_path)
        else:
            missing.append(script_path)

    return found, missing


def main() -> None:
    """Check environment integrity and warn about missing files."""
    # セッションIDの取得のためparse_hook_inputを呼び出す
    parse_hook_input()

    found, missing = check_script_files()

    if missing:
        log_hook_execution(
            "environment-integrity-check",
            "warn",
            f"Missing hook files detected: {len(missing)} files",
            {"missing_files": missing},
        )
        warning = f"""[environment-integrity-check] フック環境に問題があります

**不足ファイル** ({len(missing)}件):
{chr(10).join(f"  - {f}" for f in sorted(missing))}

**復旧方法**:
```bash
# メインリポジトリを最新に同期
git checkout main
git pull origin main

# または不足ファイルを復元
git restore .claude/hooks/ scripts/
```

**原因**: 別セッションで追加されたフックがマージされていない可能性があります。
"""
        print(warning, file=sys.stderr)
    else:
        log_hook_execution(
            "environment-integrity-check",
            "approve",
            f"All {len(found)} hook files present",
        )

    # Always continue (don't block) - just warn
    print(json.dumps({"continue": True}))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/environment_integrity_check.ts
#!/usr/bin/env bun
/**
 * フック環境の整合性をチェックする。
 *
 * Why:
 *   settings.jsonに登録されたフックファイルが存在しないと、
 *   他のフックが期待通りに動作しない。セッション開始時に
 *   不足を検出し、早期に問題を認識させる。
 *
 * What:
 *   - settings.jsonから登録済みスクリプトパスを抽出
 *   - 各ファイルの存在を確認
 *   - 不足ファイルがあれば復旧手順を含む警告を表示
 *
 * Remarks:
 *   - ブロックはせず警告のみ（他の作業は継続可能）
 *   - SessionStartで発火
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#2874: TypeScriptに移植
 */

import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";

const HOOK_NAME = "environment-integrity-check";

const HOOKS_DIR = dirname(import.meta.dir);
const PROJECT_DIR = dirname(dirname(HOOKS_DIR));
const SETTINGS_FILE = join(dirname(HOOKS_DIR), "settings.json");

/**
 * Extract all script paths from settings JSON content (relative to project root).
 * This is a pure function for easy testing.
 */
export function extractScriptPathsFromSettings(settingsContent: string): string[] {
  try {
    const settings = JSON.parse(settingsContent);
    const hooks = settings.hooks || {};
    const scriptPaths: Set<string> = new Set();

    // Pattern to match paths after $CLAUDE_PROJECT_DIR (quoted or unquoted)
    // Matches: "$CLAUDE_PROJECT_DIR"/path or $CLAUDE_PROJECT_DIR/path
    const pattern = /"?\$CLAUDE_PROJECT_DIR"?\/([^\s"]+)/g;

    for (const hookType of Object.values(hooks)) {
      if (Array.isArray(hookType)) {
        for (const entry of hookType) {
          const entryObj = entry as { hooks?: { command?: string }[] };
          for (const hook of entryObj.hooks || []) {
            const cmd = hook.command || "";
            if (!cmd) continue;

            // Find all script paths in the command
            pattern.lastIndex = 0;
            let match: RegExpExecArray | null = pattern.exec(cmd);
            while (match !== null) {
              let path = match[1];
              if (path) {
                // Clean up the path (remove trailing quotes, etc.)
                path = path.replace(/["']/g, "").trim();
                // Only include actual script files (not directories)
                if (path.endsWith(".py") || path.endsWith(".sh") || path.endsWith(".ts")) {
                  scriptPaths.add(path);
                }
              }
              match = pattern.exec(cmd);
            }
          }
        }
      }
    }

    return Array.from(scriptPaths);
  } catch {
    return [];
  }
}

/**
 * Extract all script paths from settings.json (relative to project root).
 */
export function getRegisteredScripts(): string[] {
  if (!existsSync(SETTINGS_FILE)) {
    return [];
  }

  try {
    const content = readFileSync(SETTINGS_FILE, "utf-8");
    return extractScriptPathsFromSettings(content);
  } catch {
    return [];
  }
}

/**
 * Check which script files exist and which are missing.
 */
function checkScriptFiles(): { found: string[]; missing: string[] } {
  const registered = getRegisteredScripts();
  const found: string[] = [];
  const missing: string[] = [];

  for (const scriptPath of registered) {
    const fullPath = join(PROJECT_DIR, scriptPath);
    if (existsSync(fullPath)) {
      found.push(scriptPath);
    } else {
      missing.push(scriptPath);
    }
  }

  return { found, missing };
}

async function main(): Promise<void> {
  // Parse hook input to get session ID
  const hookInput = await parseHookInput();
  const sessionId = hookInput.session_id;

  const { found, missing } = checkScriptFiles();

  if (missing.length > 0) {
    await logHookExecution(
      HOOK_NAME,
      "approve",
      `Missing hook files detected: ${missing.length} files`,
      undefined,
      { sessionId },
    );
    const fileList = missing
      .sort()
      .map((f) => `  - ${f}`)
      .join("\n");
    const warning = `[${HOOK_NAME}] フック環境に問題があります\n\n**不足ファイル** (${missing.length}件):\n${fileList}\n\n**復旧方法**:\n\`\`\`bash\n# メインリポジトリを最新に同期\ngit checkout main\ngit pull origin main\n\n# または不足ファイルを復元\ngit restore .claude/hooks/ scripts/\n\`\`\`\n\n**原因**: 別セッションで追加されたフックがマージされていない可能性があります。`;
    console.error(warning);
  } else {
    await logHookExecution(
      HOOK_NAME,
      "approve",
      `All ${found.length} hook files present`,
      undefined,
      {
        sessionId,
      },
    );
  }

  // Always continue (don't block) - just warn
  console.log(JSON.stringify({ continue: true }));
}

if (import.meta.main) {
  main();
}
